File: app/src/main/java/com/xfl/kakaotalkbot/LoggerScreen.java
package com.xfl.kakaotalkbot;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.text.Spanned;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ScrollView;
import android.widget.TextView;

/**
 * Created by XFL on 2/19/2018.
 */

public class LoggerScreen extends AppCompatActivity {
    public static TextView logTxt;
    public static ScrollView scrollView;
    private static String scriptName;
    public FloatingActionButton goDownFab;

    public static void appendLogText(final Spanned str) {
        boolean move = false;

        if (scrollView != null) {

            if (scrollView.getScrollY() >= scrollView.getChildAt(0).getMeasuredHeight() - scrollView.getMeasuredHeight()) {

                move = true;
            }
        }
        NotificationListener.UIHandler.post(new Runnable() {
            @Override
            public void run() {
                if (logTxt != null) {
                    logTxt.append(str);

                }
            }
        });

        if (move && scrollView != null) {
            scrollView.post(new Runnable() {
                @Override
                public void run() {
                    scrollView.fullScroll(View.FOCUS_DOWN);
                }
            });
        }
    }

    public static void clearLogText() {
        if (logTxt != null)
            logTxt.setText("");
    }

    public static void initialize() {
        if (scrollView != null) {
            scrollView.post(new Runnable() {
                @Override
                public void run() {
                    scrollView.setScrollY(MainApplication.getContext().getSharedPreferences("logger", 0).getInt("scrollState", 0));
                }
            });
        }
    }

    private static String getLog() {
        Context ctx = MainApplication.getContext();
        return ctx.getSharedPreferences("log", 0).getString("log", "");
    }


    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_logger);

        scriptName = getIntent().getExtras().getString("scriptName");
        logTxt = findViewById(R.id.logTxt);
        goDownFab = findViewById(R.id.fab_goDown);
        scrollView = findViewById(R.id.logger_scrollView);
        logTxt.setText(Html.fromHtml(getLog()));
        initialize();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            scrollView.setOnScrollChangeListener(new View.OnScrollChangeListener() {
                @Override
                public void onScrollChange(View view, int i, int i1, int i2, int i3) {
                    MainApplication.getContext().getSharedPreferences("logger", 0).edit().putInt("scrollState", scrollView.getScrollY()).apply();
                    if (scrollView.getScrollY() >= scrollView.getChildAt(0).getMeasuredHeight() - scrollView.getMeasuredHeight()) {

                        goDownFab.hide();
                    } else {
                        goDownFab.show();
                    }
                }
            });
        }
        goDownFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                scrollView.post(new Runnable() {
                    @Override
                    public void run() {
                        scrollView.fullScroll(View.FOCUS_DOWN);
                    }
                });
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.clear_log, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_clearlogs) {
            clearLogText();

            Log.clear();
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/me/computerpark/ara_android/LoggerScreen.kt
package me.computerpark.ara_android

import android.os.Build
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AppCompatActivity
import android.text.Html
import android.text.Spanned
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ScrollView
import android.widget.TextView

/**
 * Created by XFL, modified by 컴터박 on 2/19/2018.
 */

class LoggerScreen : AppCompatActivity() {
    lateinit var goDownFab: FloatingActionButton


    public override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_logger)

        scriptName = intent.extras!!.getString("scriptName")
        logTxt = findViewById(R.id.logTxt)
        goDownFab = findViewById(R.id.fab_goDown)
        scrollView = findViewById(R.id.logger_scrollView)
        logTxt!!.text = Html.fromHtml(log)
        initialize()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            scrollView!!.setOnScrollChangeListener { _, _, _, _, _ ->
                MainApplication.context!!.getSharedPreferences("logger", 0).edit().putInt("scrollState", scrollView!!.scrollY).apply()
                if (scrollView!!.scrollY >= scrollView!!.getChildAt(0).measuredHeight - scrollView!!.measuredHeight) {

                    goDownFab.hide()
                } else {
                    goDownFab.show()
                }
            }
        }
        goDownFab.setOnClickListener { scrollView!!.post { scrollView!!.fullScroll(View.FOCUS_DOWN) } }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.clear_log, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == R.id.action_clearlogs) {
            clearLogText()

            Log.clear()
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {

        var logTxt: TextView? = null
        var scrollView: ScrollView? = null
        private var scriptName: String? = null

        fun appendLogText(str: Spanned) {
            var move = false

            if (scrollView != null) {

                if (scrollView!!.scrollY >= scrollView!!.getChildAt(0).measuredHeight - scrollView!!.measuredHeight) {

                    move = true
                }
            }
            NotificationListener.UIHandler!!.post {
                if (logTxt != null) {
                    logTxt!!.append(str)

                }
            }

            if (move && scrollView != null) {
                scrollView!!.post { scrollView!!.fullScroll(View.FOCUS_DOWN) }
            }
        }

        fun clearLogText() {
            if (logTxt != null)
                logTxt!!.text = ""
        }

        fun initialize() {
            if (scrollView != null) {
                scrollView!!.post { scrollView!!.scrollY = MainApplication.context!!.getSharedPreferences("logger", 0).getInt("scrollState", 0) }
            }
        }

        private val log: String
            get() {
                val ctx = MainApplication.context
                return ctx!!.getSharedPreferences("log", 0).getString("log", "")
            }
    }
}
